Fix WebDAV by requesting new ETag.

icarito opened the pull request:
This fixes saving for me. Probably fixes https://github.com/Jermolene/TiddlyWiki5/issues/2995 and improves #3073 .

ETag is supposed to verify we're not overwriting a previously changed
file. That's pretty cool.

I figured this was saving only the first time subsequently failing and
revised the requests, noticing it didn't get a new ETag after saving.

In my WebDAV service (WsgiDAV) - ETag is only served from a HEAD
request.

pmario commented:
The problem is, that line 37 doesn't work with IIS Webdav ... It returns an error. ... But I need to investigate a bit more. 

I think ETag handling depends on the backend you use .... There is an other open issue somewhere.

icarito commented:
Now the only failing condition is when you are saving and the PUT operation is still ongoing and you click Save again before it finishes - in this case you'll get a 412. There should be a way to not be able to save when a save operation is ongoing , but this is minor imho.

pmario commented:
With this code I do "also" get `406 Not Acceptable` for the HEAD request with IIS

IMO it seems to be a problem with the "accept" header, which is `application/json` It seems IIS doesn't want it that way.

pmario commented:
If I change the HEAD request like this: 

```
/*
Retrieve ETag if available
*/
var RetrieveETag = function(self) {
	var headers = { "Content-Type": "*;charset=UTF-8" };
	$tw.utils.httpRequest({
		url: self.uri(),
		type: "HEAD",
		headers: headers,
		callback: function(err, data, xhr) {
			if(!err) {
				self.etag = xhr.getResponseHeader("ETag");
			}
		}
	});
};
```

It reads the ETAG with 304 ... not modified ... Which is OK

The problem then is: If I save it ... I get 409 conflict

But the "if-match" header is right. ....

icarito commented:
That is strange! Because you shouldn't be calling `RetrieveETag` if your webserver already provides an ETag on PUT (which is what I understood - in fact [I patched WsgiDAV](https://github.com/mar10/wsgidav/pull/100) to do just this as I thought it made a lot of sense. The version that is now in the repo works without calling HEAD after PUT. So I made it such that if PUT provides an ETag then no HEAD request should be done in the first place.

It's difficult to support proprietary DAV implementations because I can't test it.

icarito commented:
@pmario hey I saw your demo of DAT syncadaptor in youtube, pretty cool!

Please confirm which HEAD request causes errors for you - I'd expect you should only make one HEAD request if no ETag is found.

pmario commented:
@icarito see: https://stackoverflow.com/questions/42246577/why-responses-to-put-requests-must-not-provide-an-etag

pmario commented:
I think I did find the problems with my setup. 

1st: HEAD request for IIS needs an `Accept: */*` otherwise it returns an error. 
2nd: I did also play around with static content compression. Which seems to cause a problem. 
   - I did deactivate compression to get consistent Etags
3rd: After the first save is successful, IIS returns a weak Etag: W/"..", which isn't allowed to be used for PUT
   - So I did add a little `.replace(/^W\//,"")` function 

and now everything seems to work nicely.

pmario commented:
can you try this code with your setup: https://github.com/pmario/TiddlyWiki5/blob/fix-3231/core/modules/savers/put.js ??

Jermolene commented:
Thanks @icarito much appreciated. Is this ready to merge? We do also need a CLA signature I'm afraid, apologies.

pmario commented:
@Jermolene ... not ready atm!

icarito commented:
@pmario I can confirm that the code you asked me to test works fine for me! Thanks for your collaboration! :-D :tada: :heart:

pmario commented:
I think my code changes aren't pushed to this PR yet? @icarito

pmario commented:
I did test the code from my branch: https://github.com/pmario/TiddlyWiki5/blob/fix-3231/core/modules/savers/put.js  with a unix nginx webdav setting. 

It seems to work!

@icarito ... If you are fine with the changes, just use them and finish the PR, so Jeremy can merge it.

pmario commented:
I did find a second docker setting, which uses an alpine OS apache webdav setup. ... But I didn't test it yet :)

pmario commented:
That's strange. I just tested again with IIS and now I get an error again. ..-> I'm testing

pmario commented:
HMMMM ... 

 - If the URL is: `http://localhost/WebDAV/` I do get a 409
 - If URL is http://localhost/WebDAV/index.html ... ---> OK 

IMO we should fix this. .. Can we fix this?

pmario commented:
>IMO we should fix this. .. Can we fix this?

IMO can't be fixed client side. .. So the server may do a redirect

pmario commented:
@Jermolene ... From my point of view it works now with my settings. win10 IIS10 and ubuntu with nginx webdav

Jermolene commented:
Great, thanks @icarito @pmario

icarito commented:
On 05/05/18 06:51, FND wrote:
>
> *@FND* commented on this pull request.
>
> A few observations on making the code more readable and robust.
>
Thanks!
> ------------------------------------------------------------------------
>
> In core/modules/savers/put.js
> <https://github.com/Jermolene/TiddlyWiki5/pull/3230#discussion_r186258878>:
>
> > @@ -15,6 +15,24 @@ to the current URL, such as a WebDAV server.
>  /*global $tw: false */
>  "use strict";
>  
> +/*
> +Retrieve ETag if available
> +*/
> +var RetrieveETag = function(self) {
> +	var headers = { "Accept": "*/*;charset=UTF-8" };
> +	$tw.utils.httpRequest({
> +		url: self.uri(),
> +		type: "HEAD",
> +		headers: headers,
> +		callback: function(err, data, xhr) {
> +			if(!err) {
> +				self.etag = xhr.getResponseHeader("ETag").replace(/^W\//,"");
>
> @pmario <https://github.com/pmario> explained
> <https://github.com/Jermolene/TiddlyWiki5/pull/3073#issuecomment-386797166>
> that modiying the ETag is necessary to remove the |W/| prefix
> indicating a weak ETag.
>
> Since that's not at all obvious from the code, we should add a comment
> to that effect here.
>
I totally agree, I had asked @pmario to explain and now that he has, we
can add a comment!

> ------------------------------------------------------------------------
>
> In core/modules/savers/put.js
> <https://github.com/Jermolene/TiddlyWiki5/pull/3230#discussion_r186258894>:
>
> > @@ -15,6 +15,24 @@ to the current URL, such as a WebDAV server.
>  /*global $tw: false */
>  "use strict";
>  
> +/*
> +Retrieve ETag if available
> +*/
> +var RetrieveETag = function(self) {
> +	var headers = { "Accept": "*/*;charset=UTF-8" };
>
> As stated elsewhere
> <https://github.com/Jermolene/TiddlyWiki5/pull/3073#issuecomment-386799601>,
> I believe we don't actually need an |Accept| header here: Since |HEAD|
> requests should never contain a response body, the content type should
> not make any difference.
>
> So we just need to prevent TiddlyWiki's |httpRequest| from defaulting
> to JSON by passing |headers: {}| below.
>

According to @pmario, this is needed in IIS so another comment should be
added mentioning it.

> ------------------------------------------------------------------------
>
> In core/modules/savers/put.js
> <https://github.com/Jermolene/TiddlyWiki5/pull/3230#discussion_r186258907>:
>
> > @@ -69,15 +78,20 @@ PutSaver.prototype.save = function(text, method, callback) {
>  		data: text,
>  		callback: function(err, data, xhr) {
>  			if(err) {
> -				callback(err);
> -			} else if(xhr.status === 200 || xhr.status === 201) {
> +				// response is textual: "XMLHttpRequest error code: 412"
> +				const status = Number(err.substring(err.indexOf(':') + 2, err.length))
>
> Why not just use |xhr.status| here? This seems very brittle.
>
> (Also because IIRC, using |Number| to parse strings is unreliable
> across browsers; |parseInt| is preferable.)
>

This is part of what I fixed. In my tests, there was no `xhr.status`
when the request errored out (i.e. was not completed) - this is with
WsgiDAV.
We can change to parseInt - I didn't know it was preferred to Number.

Regards!
Sebastian

pmario commented on core/modules/savers/put.js:
```diff
@@ -15,6 +15,24 @@ to the current URL, such as a WebDAV server.
+Retrieve ETag if available
+*/
+var RetrieveETag = function(self) {
+	var headers = { "Accept": "*/*;charset=UTF-8" };
```
needs to be tested.

pmario replied:
You are right. If we use `var headers = {}` my FF browser uses `Accept: */*` by default. ... The `charset=` is probably a copy paste problem from `.save()`

pmario replied:
So if you want to make it more readable we should stay with the accept header. Otherwise users may be curious why it is there at all. ...  

	var headers = { "Accept": "*/* };
